### PR TITLE
feat: split consultant display name into public and internal names (#996)

### PR DESCRIPTION
**Pull request by Storypapst**
Closes OpenResilienceInitiative/ORISO-UserService#996

**Stacked on #1007; retarget to `pre-dev` after it merges.**

## What

- `displayName` stays the PUBLIC name (what advice seekers see); new nullable `internal_display_name` column (Liquibase changeset `0084-consultant-internal-display-name`).
- **One central fallback rule**: `Consultant#getInternalDisplayNameOrFallback()` resolves `internalDisplayName ?? displayName` (blank counts as absent). Wired into the internal surfaces served by this service:
  - tenant team list (`GET /users/consultants/tenant`, used for internal colleague pickers)
  - case-handover candidate list (names shown to colleagues)
- Deliberately **not** rewired (audience includes advice seekers, public name must stay): Matrix room display names (ADR-002 pseudonymity), client-facing rename/supervisor notifications, group chat participant resolution.
- Admin API: `internalDisplayName` on `ConsultantDTO`/create/update (+ `displayName` now writable via the admin update, both with null-untouched / blank-clears semantics); search results carry `internalDisplayName`.
- Dead scaffolding fixed: consultant get-by-id responses now populate `publicName` as the alias of the public display name (was hardcoded `null` after enrichment); `AdminDtoMapper`/`AdminResponseDTOBuilder` populate admin `publicName` with the admin's full name (username as last resort) instead of `null`. `ConsultantDtoMapper` keeps `publicName` aliased to `displayName` — verified still correct.

## Why

The wizard design (`AvatarNameCard`) models a public name for advice seekers and an internal name with fallback. This makes both names real instead of adding a third concept. Companion Admin-panel PR: OpenResilienceInitiative/ORISO-Admin#733.

## Tests

- `JAVA_HOME=$(/usr/libexec/java_home -v 21) ./mvnw test` (full suite) → **3844 tests, 0 failures, BUILD SUCCESS**
- New coverage: entity fallback rule (set / null / blank / neither), update IT for independent persistence + clearing, tenant team list prefers the internal name, response builder + search mapper carry `internalDisplayName`, `publicName` population for consultants and admins.

## How should I test this?

- [ ] As platform admin, set public name "Anna B." and internal name "Anna Beispiel (Standort Nord)" via `PUT /useradmin/consultants/{id}` → `GET` returns both, and `publicName` equals "Anna B." (not null).
- [ ] `GET /users/consultants/tenant` as a consultant of the same tenant → the colleague appears as "Anna Beispiel (Standort Nord)".
- [ ] As an advice seeker talking to that consultant → still sees "Anna B." everywhere (session list, chat header, notifications).
- [ ] Clear the internal name (`"internalDisplayName": ""`) → team list falls back to "Anna B.".
- [ ] `PUT` without either name field → both stored names stay unchanged.
- [ ] `GET /useradmin/tenantadmins/{id}` (or agency admin) → `publicName` is the admin's full name, not null.
- [ ] Case handover: search handover candidates as a counsellor → candidate consultant names use the internal name with fallback.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **New Features**
  - Added dedicated counsellor invitation onboarding with profile setup, topic selection, registration, and two-factor authentication.
  - Counsellor invitations now open the dedicated onboarding experience, including resumable 2FA setup.
  - Added separate public and internal consultant display names with automatic fallback.
  - Consultant searches and handover candidates now include internal display names.
  - Administrator public names now use personal names, with username fallback.

- **Bug Fixes**
  - Improved handling of blank or unavailable display names and expired invitations.

- **Tests**
  - Added comprehensive coverage for onboarding, 2FA, display names, search, and persistence.

### CODE REVIEW

**Review by shazia-k — COMMENTED** [review]
@Storypapst there are 2 major findings by code rabbit which needs to address please

File: src/main/java/de/caritas/cob/userservice/api/admin/service/consultant/ConsultantAdminService.java [hunk 1]
```diff
@@ -121,7 +121,14 @@ private void enrichWithTopics(String consultantId, ConsultantAdminResponseDTO re
...
         .findConsultant(consultantId)
-        .ifPresent(map -> response.getEmbedded().setDisplayName(getDisplayNameFromUserMap(map)));
+        .ifPresent(
+            map -> {
+              var displayName = getDisplayNameFromUserMap(map);
+              response.getEmbedded().setDisplayName(displayName);
+              // publicName is an alias of the PUBLIC display name (#996) — it used to be
+              // dead scaffolding hardcoded to null.
+              response.getEmbedded().setPublicName(displayName);
+            });
```
[the remark this answers is not in the document]
Comment (Storypapst) [reply src/main/java/de/caritas/cob/userservice/api/admin/service/consultant/ConsultantAdminService.java:131]: Applied: `createNewConsultant` and `updateConsultant` now call the same `enrichWithDisplayName` as `findConsultantById`, so all three admin response paths carry displayName + publicName; unit tests added for both paths.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/main/java/de/caritas/cob/userservice/api/service/CaseHandoverService.java [hunk 2]
```diff
@@ -501,7 +501,8 @@ private ConsultantSessionResponseDTO toCandidateDto(Session session) {
               .firstName(consultant.getFirstName())
               .lastName(consultant.getLastName())
               .username(decodeUsername(consultant.getUsername()))
-              .displayName(decodeUsername(consultant.getDisplayName())));
+              // Handover candidates are shown to colleagues (internal surface, #996).
+              .displayName(decodeUsername(consultant.getInternalDisplayNameOrFallback())));
```
[the remark this answers is not in the document]
Comment (Storypapst) [reply src/main/java/de/caritas/cob/userservice/api/service/CaseHandoverService.java:512]: Applied: the candidate filter now also searches `getInternalDisplayNameOrFallback()` (public display name kept as an additional term), matching what the list renders; regression test `searchCandidates_matchesInternalDisplayNameOnlyQuery` covers an internal-name-only query.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/main/java/de/caritas/cob/userservice/api/UserServiceMapper.java [hunk 3]
```diff
@@ -287,6 +287,9 @@ public Map<String, Object> mapOf(
     map.put("isTeamConsultant", nonNull(fullConsultant) && fullConsultant.isTeamConsultant());
     map.put("isSupervisor", nonNull(fullConsultant) && fullConsultant.isSupervisor());
     map.put("displayName", nonNull(fullConsultant) ? fullConsultant.getDisplayName() : null);
+    map.put(
+        "internalDisplayName",
+        nonNull(fullConsultant) ? fullConsultant.getInternalDisplayName() : null);
```
[the remark this answers is not in the document]
Comment (Storypapst) [reply src/main/java/de/caritas/cob/userservice/api/UserServiceMapper.java:292]: Applied: both `findAllByInfix` and `findAllByInfixAndAgencyIds` now match `internalDisplayName`; two repository ITs assert the delta for an internal-name-only infix.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/test/java/de/caritas/cob/userservice/api/adapters/web/mapping/AdminDtoMapperTest.java [hunk 4]
```diff
@@ -38,7 +38,8 @@ void adminSearchResultOf_Should_NotFail_WhenTenantServiceReturnsNotFound() {
     assertThat(result.getEmbedded().get(0).getEmbedded().getTenantId()).isEqualTo("2");
     assertThat(result.getEmbedded().get(0).getEmbedded().getTenantName()).isNull();
     assertThat(result.getEmbedded().get(0).getEmbedded().getTenantSubdomain()).isNull();
-    assertThat(result.getEmbedded().get(0).getEmbedded().getPublicName()).isNull();
+    // publicName used to be dead scaffolding hardcoded to null (#996) — it is the full name now.
+    assertThat(result.getEmbedded().get(0).getEmbedded().getPublicName()).isEqualTo("First Last");
```
[the remark this answers is not in the document]
Comment (Storypapst) [reply src/test/java/de/caritas/cob/userservice/api/adapters/web/mapping/AdminDtoMapperTest.java:42]: Applied: added `adminSearchResultOf_Should_FallBackToUsernameForPublicName_When_NamesAreBlank` asserting publicName equals the username when first/last name are blank.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/test/java/de/caritas/cob/userservice/api/adapters/web/mapping/ConsultantDtoMapperTest.java [hunk 5]
```diff
@@ -62,6 +62,7 @@ void consultantDtoOf_Should_MapMasterTableFields() {
     // then
     assertThat(consultant.getDisplayName()).isEqualTo("Public Name");
     assertThat(consultant.getPublicName()).isEqualTo("Public Name");
+    assertThat(consultant.getInternalDisplayName()).isEqualTo("Internal Name");
```
[the remark this answers is not in the document]
Comment (Storypapst) [reply src/test/java/de/caritas/cob/userservice/api/adapters/web/mapping/ConsultantDtoMapperTest.java:65]: Applied: added a case with no `internalDisplayName` in the map — the DTO keeps it null while displayName and publicName stay the public name.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by Storypapst** [comment]
@shazia-k Both major findings are fixed and pushed (`9ccab828`), plus the three minor ones:

- **ConsultantAdminService** — `createNewConsultant` and `updateConsultant` now run the same `enrichWithDisplayName` as `findConsultantById`, so create, update and get all expose the same public-name contract.
- **CaseHandoverService** — the candidate filter now searches the internal display name with fallback (the public name stays a valid search term), so an internal-name-only query no longer drops the session before the list renders it.
- Minor: both consultant infix searches cover `internalDisplayName`; added tests for the username fallback of `publicName` and for a null internal name in the DTO mapper.

New tests: both admin response paths, two repository infix ITs, and `searchCandidates_matchesInternalDisplayNameOnlyQuery`. Full suite green locally: 3867 tests, 0 failures. All five review threads answered and resolved — ready for another look.

**Comment by Storypapst** [comment]
@coderabbitai review

All five findings from the previous review are fixed in `9ccab828` and every thread is resolved — requesting a fresh review so the stale CHANGES_REQUESTED state reflects the current head.

File: src/main/java/de/caritas/cob/userservice/api/adapters/web/controller/TenantAdminOnboardingController.java [hunk 6]
```diff
@@ -62,10 +88,41 @@ public ResponseEntity<TenantAdminRegistrationResponseDTO> registerTenantAdmin(
...
     return ResponseEntity.ok().build();
   }
 
+  /**
+   * Role probe deciding which onboarding flow answers (unknown tokens 404 here already). The
+   * role-specific services re-validate the role themselves, so a mismatch can never slip through.
+   */
+  private AccountInviteTargetRole targetRoleOf(String token) {
+    return accountInviteService.findInviteByToken(token).getTargetRole();
+  }
```
[the remark this answers is not in the document]
Comment (Storypapst) [reply src/main/java/de/caritas/cob/userservice/api/adapters/web/controller/TenantAdminOnboardingController.java:109]: Fixed in f5baaaf6 — the dispatch probe now projects the invite's target role alone through a lock-free query, so the row is lock-requested once per request (by the flow that actually works on it) instead of twice; unknown/blank tokens still answer 404/400 and the services still re-validate the role against the row they locked.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/main/java/de/caritas/cob/userservice/api/service/accountinvite/onboarding/CounsellorOnboardingService.java [hunk 7]
```diff
@@ -0,0 +1,454 @@
...
+    if (invite.getStatus() == AccountInviteStatus.EMAIL_SENT) {
+      expireIfPastExpiry(invite, now);
+      return new CounsellorOnboardingState(invite, false, resolveTopicCoverage(invite).topics());
+    }
+    if (isResumableAtTwoFactorStep(invite, now)) {
+      repairMissingTotpSecret(invite);
+      return new CounsellorOnboardingState(invite, true, List.of());
+    }
+    throw linkDeathException(invite);
+  }
```
[the remark this answers is not in the document]
Comment (Storypapst) [reply src/main/java/de/caritas/cob/userservice/api/service/accountinvite/onboarding/CounsellorOnboardingService.java]: Fixed in e770dec5 — the transactional scope of both onboarding flows is now database-only (locked load, expiry transition, state classification, gate consumption) and every remote call runs outside it: topic coverage after the resolve transaction commits, and the Keycloak round trips of activateTwoFactor/repairMissingTotpSecret between two short transactions. The EXPIRED-before-link-death guarantee is preserved structurally (the transactional step returns the link-death exception, the caller throws it after the commit) instead of via noRollbackFor, no REQUIRES_NEW on the locked row is used, and registerTenantAdmin keeps its annotated atomic transaction. The same treatment was applied to the identical pattern in TenantAdminOnboardingService (DPA text lookup and 2FA activation).

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/main/java/de/caritas/cob/userservice/api/service/accountinvite/onboarding/CounsellorOnboardingService.java [hunk 8]
```diff
@@ -0,0 +1,454 @@
...
+      return;
+    }
+    try {
+      var profile = identityProfileLookup.findById(invite.getAcceptedByUserId()).orElse(null);
+      if (profile == null) {
+        log.warn(
+            "Cannot repair the missing TOTP setup material for invite {} — no identity profile"
+                + " exists for user {}",
+            invite.getId(),
+            invite.getAcceptedByUserId());
```
[the remark this answers is not in the document]
Comment (Storypapst) [reply src/main/java/de/caritas/cob/userservice/api/service/accountinvite/onboarding/CounsellorOnboardingService.java]: Fixed in e6b8d372 — the identity-provider user id is gone from the warning, the invite id stays; a sweep over the wizard code found no other log statement carrying user ids, usernames, e-mails, tokens or Matrix ids.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/main/java/de/caritas/cob/userservice/api/service/accountinvite/onboarding/TenantAdminOnboardingService.java [hunk 9]
```diff
@@ -71,7 +71,9 @@ public class TenantAdminOnboardingService {
    * must be on screen and navigable (anchor/TOC) rather than replaced by a placeholder hint. The
    * resume path skips the lookup — it re-enters at the 2FA step, which shows no contract.
    */
-  @Transactional
+  // noRollbackFor: expireIfPastExpiry persists the EXPIRED transition and then throws the
+  // link-death exception — without it the rollback would keep the row EMAIL_SENT forever.
+  @Transactional(noRollbackFor = AccountInviteLinkException.class)
```
[the remark this answers is not in the document]
Comment (Storypapst) [reply src/main/java/de/caritas/cob/userservice/api/service/accountinvite/onboarding/TenantAdminOnboardingService.java]: Fixed in 50811c47 — a Spring-backed IT now calls the real services, catches the link-death exception and reads the invite back in a separate transaction, for tenant-admin resolve + register and the counsellor equivalents; verified it can fail by dropping noRollbackFor from registerTenantAdmin (register test red) and by throwing the expiry inside the transactional step (both resolve tests red), then restored.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/test/java/de/caritas/cob/userservice/api/adapters/web/controller/CounsellorOnboardingWizardIT.java [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+                      "topicIds": [999]
+                    }
+                    """))
+        .andExpect(status().isInternalServerError());
+
+    mockMvc
+        .perform(get("/users/account-invites/{token}/onboarding", token))
+        .andExpect(status().isOk())
+        .andExpect(jsonPath("$.phase").doesNotExist());
+  }
```
[the remark this answers is not in the document]
Comment (Storypapst) [reply src/test/java/de/caritas/cob/userservice/api/adapters/web/controller/CounsellorOnboardingWizardIT.java]: Fixed in 70bf6849 — AgencyService is a @MockitoBean now: a healthy stub with a known topic set pins the documented 400 branch and the resolved coverage asserted in the flow test, a throwing stub pins the degraded 500 branch; no assertion depends on the upstream being unreachable in the testing profile any more.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by Storypapst** [comment]
@shazia-k Thanks for merging #1020 — since the wizard code landed on this branch, CodeRabbit reviewed it for the first time and raised five more findings. All are fixed and pushed (head `70bf6849`), suite green (**3899 tests, 0 failures**), and **all 10 review threads are resolved**:

- **Privacy (major)** — a WARN log wrote the invitee's identity-provider user id; removed, invite id kept for support correlation. Swept the rest of the wizard/invite code for further identifiers in logs: none found.
- **Lock contention (major)** — `resolveOnboardingInvite` and `activateTwoFactor` held the invite's `PESSIMISTIC_WRITE` lock while calling AgencyService, TopicService and Keycloak. The transactional step now covers DB work only; remote calls run after commit. The `noRollbackFor` guarantee became structural: the transactional step *returns* the link-death exception and the caller throws it after the EXPIRED transition has committed. Same treatment applied to the identical pattern in `TenantAdminOnboardingService`.
- **Duplicate row lock** — the role probe now uses a lock-free projection, so the invite row is lock-requested exactly once per request.
- **Test that could not fail** — new `OnboardingInviteExpiryCommitIT` reads the invite back in a separate transaction and asserts the committed EXPIRED state; verified it goes red when `noRollbackFor` is removed or the throw is moved back inside the transaction.
- **Test asserting an environmental accident** — `AgencyService` is now a `@MockitoBean`; the 400 (healthy coverage) and 500 (agency down) branches are driven explicitly instead of relying on the upstream being unreachable in the testing profile.

One thing left deliberately unchanged: `registerTenantAdmin` still calls Keycloak and TenantService inside its transaction — that is its intended atomic unit with explicit compensation, so re-cutting those boundaries felt riskier than the review round warranted. Happy to open a separate ticket if you'd like it addressed.

Ready for another look.

File: src/main/java/de/caritas/cob/userservice/api/service/accountinvite/onboarding/TenantAdminOnboardingService.java [hunk 11]
```diff
@@ -231,10 +259,55 @@ public void activateTwoFactor(String rawToken, String oneTimePassword) {
...
+          accountInviteRepository.save(invite);
+          accountInviteService.markTwoFactorActive(invite.getAcceptedByUserId());
+          return null;
+        });
+  }
+
+  /** Runs {@code action} in its own short database transaction (no remote call belongs inside). */
+  private <T> T inTransaction(Supplier<T> action) {
+    return new TransactionTemplate(transactionManager).execute(status -> action.get());
   }
```
[the remark this answers is not in the document]
Comment (Storypapst) [reply src/main/java/de/caritas/cob/userservice/api/service/accountinvite/onboarding/TenantAdminOnboardingService.java:311]: Done in ba34bb57: both onboarding services now build one reusable `TransactionTemplate` from the injected `PlatformTransactionManager` in their constructor instead of allocating one per `inTransaction` call.
_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/test/java/de/caritas/cob/userservice/api/service/accountinvite/onboarding/OnboardingInviteExpiryCommitIT.java [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+                .departmentId(2L)
+                .tokenHash(AccountInviteService.hash(rawToken))
+                .expiresAt(LocalDateTime.now().minusDays(1))
+                .status(AccountInviteStatus.EMAIL_SENT)
+                .emailVerificationStatus(EmailVerificationStatus.PENDING)
+                .twoFactorStatus(TwoFactorGateStatus.PENDING_SETUP)
+                .createDate(LocalDateTime.now().minusDays(8))
+                .build())
+        .getId();
+  }
```
[the remark this answers is not in the document]
Comment (Storypapst) [reply src/test/java/de/caritas/cob/userservice/api/service/accountinvite/onboarding/OnboardingInviteExpiryCommitIT.java:142]: Done in 9f52463c: `OnboardingInviteExpiryCommitIT` now tracks its seeded invite ids and deletes them in an `@AfterEach`, so the real-database run no longer leaves rows behind (the `AutoConfigureTestDatabase` import stays as it is).
_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by Storypapst** [comment]
Follow-up round for the two findings from the review's outside-diff section. Both are addressed on `feat/consultant-dual-name-996` (`27fbe2bf`), full suite green (3899 tests). Heads-up: this PR was already merged into `pre-dev` while the round was running, so the four commits sit on the re-created branch and still need a follow-up PR to land.

**MAJOR — detached `AccountInvite` snapshots could lose a concurrent update (`CounsellorOnboardingService`)** — fixed in `39fdd364`. `AccountInvite` has neither `@Version` nor `@DynamicUpdate`, so saving the snapshot a flow carried across a remote call rewrote every column with its pre-call value. Each such write now re-reads the row under its `PESSIMISTIC_WRITE` lock inside a short transaction (new `AccountInviteRepository#findByIdForUpdate`) and sets only the fields it owns:

- `registerCounsellor` — the precondition check plus the `EMAIL_SENT -> EXPIRED` transition moved into one short transaction that still RETURNS the link death (so the transition commits before the exception is thrown), and the pending TOTP secret is written into a row re-read after the Keycloak round trip instead of into the snapshot the provisioning returned.
- `repairMissingTotpSecret` — the repaired secret is persisted through the same targeted update; the in-memory copy is updated too because the resume response is answered from it.
- `consumeTwoFactorGate` in **both** services — the row is re-read under its lock, and clearing the secret stays FIRST so `markTwoFactorActive`, which re-reads the invite by its acceptor and therefore works on the very same managed row, is not undone.

Agency/Topic/Keycloak/DPA calls all stay outside the locked transactions. `registerTenantAdmin` is deliberately left alone: its single atomic transaction is what keeps tenant creation, Keycloak account and the irreversible reservation consumption compensable together (it has its own `rollbackUser` compensation), and every invite it writes is a managed entity re-read inside that very transaction — no detached merge there.

New `OnboardingInviteTargetedUpdateIT` proves it against the real database: a concurrent rename of an unrelated field commits in its own transaction *inside* the stubbed Keycloak call — exactly between the flow's read and its write — and the test asserts both that the owned field was written and that the concurrent change survived. Revert check: restoring `save(detachedSnapshot)` in the two write paths turned both tests red on exactly that assertion (`expected: "Concurrently Renamed" but was: "Simpson"`); restoring the fix turned them green again.

**MINOR — `CounsellorOnboardingWizardIT` never proved the agency-only topic** — fixed in `27fbe2bf`. The successful flow now submits the agency-only topic `7` next to the routed department topic `2` and asserts both reach the provisioning command, so a regression that offers topic 7 on resolve but drops or rejects it on register fails the test.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Review by shazia-k — COMMENTED** [review]
Looks good
